Add particle iterator and accessor

gassmoeller opened the pull request:
This PR is not  quite complete, but I would like to ask for a review by @bangerth to see if I got the concept right. I added `ParticleIterator` and `ParticleAccessor` classes and converted one of the direct uses inside of `World` to use the iterators to test they work. The intention is to convert all direct uses of particles into using the iterator class.

bangerth commented on include/aspect/particle/particle_iterator.h:
```diff
@@ -0,0 +1,60 @@
+      ParticleAccessor<dim,spacedim> accessor;
+    public:
+
+      ParticleIterator () = default;
```
wouldn't this require a default constructor of the accessor class?

bangerth commented on include/aspect/particle/world.h:
```diff
@@ -74,6 +76,26 @@ namespace aspect
         /**
+         * Return an iterator to the first particle.
+         */
+        ParticleIterator<dim> begin() const;
```
Can you introduce a typedef
```
  typedef ParticleIterator<dim> particle_iterator;
```
(or just `iterator` if you want)? This is how containers commonly do things.

bangerth commented on source/particle/world.cc:
```diff
@@ -192,11 +220,8 @@ namespace aspect
-          locally_highest_index = std::max(locally_highest_index,it->second.get_id());
-        }
+      for(ParticleIterator<dim> particle = begin(); particle != end(); ++particle)
+        locally_highest_index = std::max(locally_highest_index,particle->get_id());
```
OK, so it already works :-)

gassmoeller commented:
I documented the new functions and addressed your comments except for the `surrounding_cell` function. I would delay this for the next PR, when I start creating a `ParticleHandler` and separate Particle and ASPECT related funtionality.
This PR is ready from my point of view.

bangerth commented on include/aspect/particle/particle_accessor.h:
```diff
@@ -0,0 +1,188 @@
+        /**
+         * Make ParticleIterator a friend to allow it constructing ParticleAccessors.
+         */
+        template <int, int> friend class ParticleIterator;
```
I think we always move the friend declarations to the very end of the class. Can you do the same here?

bangerth commented on include/aspect/particle/particle_iterator.h:
```diff
@@ -0,0 +1,95 @@
+    template<int dim, int spacedim=dim>
+    class ParticleIterator
+    {
+      private:
```
move the `private` section below the `public` section for consistency with other places of the code base.

bangerth commented on include/aspect/particle/particle_iterator.h:
```diff
@@ -0,0 +1,95 @@
+        /**
+         * Increment operator. Advance to the next particle.
+         */
+        void operator++();
```
You'll want to have both the prefix and postfix versions of these operators, so that people can write `++particle` and `particle++` and have both do the same thing.

bangerth commented on source/particle/particle_accessor.cc:
```diff
@@ -0,0 +1,189 @@
+      particle (particle)
+    {}
+
+    template <int dim, int spacedim>
```
two more empty lines

bangerth commented on source/particle/particle_accessor.cc:
```diff
@@ -0,0 +1,189 @@
+      return particle->second.get_properties();
+    }
+
+    template <int dim, int spacedim>
```
two more empty lines. same below

gassmoeller commented:
I addressed all comments.

bangerth commented:
See #1897.
